src/Business Foundation/App/NoSeriesCopilot/src/Copilot/NoSeriesCopilotImpl.Codeunit.al

```diff
@@ -0,0 +1,509 @@
+    var
+        JsonObject: JsonObject;
+        JsonArrayToken: JsonToken;
+        XPathLbl: Label '$.noSeries', Locked = true;
+    begin
+        JsonObject.ReadFrom(Completion);
+        JsonObject.SelectToken(XPathLbl, JsonArrayToken);
+        exit(JsonArrayToken.AsArray());
+    end;
+
+    local procedure SaveGenerationHistory(var NoSeriesProposal: Record "No. Series Proposal"; InputText: Text)
+    begin
+        NoSeriesProposal.Init();
+        NoSeriesProposal."No." := NoSeriesProposal.Count + 1;
+        NoSeriesProposal.SetInputText(InputText);
+        NoSeriesProposal.Insert(true);
+    end;
+
+    local procedure CreateNoSeries(var NoSeriesProposal: Record "No. Series Proposal"; var NoSeriesGenerated: Record "No. Series Proposal Line"; Completion: Text)
+    var
+        Json: Codeunit Json;
+        NoSeriesArrText: Text;
+        NoSeriesObj: Text;
+        i: Integer;
+    begin
+        ReadGeneratedNumberSeriesJArray(Completion).WriteTo(NoSeriesArrText);
+        ReAssambleDuplicates(NoSeriesArrText);
+
+        Json.InitializeCollection(NoSeriesArrText);
+
+        for i := 0 to Json.GetCollectionCount() - 1 do begin
+            Json.GetObjectFromCollectionByIndex(i, NoSeriesObj);
+
+            InsertNoSeriesGenerated(NoSeriesGenerated, NoSeriesObj, NoSeriesProposal."No.");
+        end;
+    end;
+
+    local procedure ReAssambleDuplicates(var NoSeriesArrText: Text)
+    var
+        i: Integer;
+        NoSeriesObj: Text;
+        NoSeriesCodes: List of [Text];
+        NoSeriesCode: Text;
+        Json: Codeunit Json;
+    begin
+        Json.InitializeCollection(NoSeriesArrText);
+
+        for i := 0 to Json.GetCollectionCount() - 1 do begin
+            Json.GetObjectFromCollectionByIndex(i, NoSeriesObj);
+            Json.InitializeObject(NoSeriesObj);
+            Json.GetStringPropertyValueByName('seriesCode', NoSeriesCode);
+            if NoSeriesCodes.Contains(NoSeriesCode) then begin
+                Json.ReplaceOrAddJPropertyInJObject('seriesCode', GenerateNewSeriesCodeValue(NoSeriesCodes, NoSeriesCode));
+                NoSeriesObj := Json.GetObjectAsText();
+                Json.ReplaceJObjectInCollection(i, NoSeriesObj);
+            end;
+            NoSeriesCodes.Add(NoSeriesCode);
+        end;
+
+        NoSeriesArrText := Json.GetCollectionAsText()
+    end;
+
+    local procedure GenerateNewSeriesCodeValue(var NoSeriesCodes: List of [Text]; var NoSeriesCode: Text): Text
+    var
+        NewNoSeriesCode: Text;
+    begin
+        repeat
+            NewNoSeriesCode := CopyStr(NoSeriesCode, 1, 18) + '-' + RandomCharacter();
+        until not NoSeriesCodes.Contains(NewNoSeriesCode);
+
+        NoSeriesCode := NewNoSeriesCode;
+        exit(NewNoSeriesCode);
+    end;
+
+    local procedure RandomCharacter(): Char
+    begin
+        exit(RandIntInRange(33, 126)); // ASCII: ! (33) to ~ (126)
+    end;
+
+    local procedure RandIntInRange("Min": Integer; "Max": Integer): Integer
```

Comment:
Max and Min parameters with " and than used without ". Should be consistend, also matches keywork, could be changed to MinInt, MaxInt.